pulp_python/app/pypi.py

```diff
@@ -0,0 +1,133 @@
+        """
+        Upload a package to the index.
+
+        1. Check if the index allows uploaded packages (live-api-enabled)
+        2. Check if the package is in the repository already
+        3. If present then reject request
+        4. Spawn task to add content if no/old session present
+        5. Add uploads to current session to group into one task
+        """
+        def create_upload_task(cur_session):
+            cur_session['start'] = str(start_time)
+            cur_session['artifacts'] = [(str(artifact.pk), file.name)]
+            cur_session.modified = False
+            cur_session.save()
+            result = dispatch(tasks.upload, [artifact, repo],
+                                   kwargs={"session_pk": str(cur_session.session_key),
+                                           "repository_pk": str(repo.pk)
+                                           })
+            log.info(f"Dispatched task to add package {file.name}: {reverse('tasks-detail', args=[result.pk], request=None)}")
+        live_enabled = True
+        distro = self.get_distribution(path)
+        if live_enabled:
+            repo = distro.repository
+            if not repo:
+                return HttpResponseBadRequest("Index is not pointing to a repository")
+
+            repo_content = self.get_content(self.get_repository_version(distro))
+            if repo_content.filter(filename=file.name).exists():
+                return HttpResponseBadRequest(reason=f"Package {file.name} already exists in index")
+
+            artifact = Artifact.init_and_validate(file, expected_digests=metadata["digests"], expected_size=file.size)
+            try:
+                artifact.save()
+            except IntegrityError:
+                artifact = Artifact.objects.get(sha256=artifact.sha256)
+                log.info("Artifact already existed in database")
+
+            session = metadata["session"]
+            start_time = datetime.now(tz=timezone.utc) + timedelta(seconds=5)
+            if not session.get("start"):
+                create_upload_task(session)
+            else:
+                sq = Session.objects.select_for_update(nowait=False).filter(pk=session.session_key)
+                with transaction.atomic():
+                    try:
+                        sq.first()
+                        if datetime.fromisoformat(session['start']) >= datetime.now(tz=timezone.utc):
+                            session['artifacts'].append((str(artifact.pk), file.name))
+                            session['start'] = str(start_time)
+                            session.modified = False
+                            session.save()
+                            log.info(f"Updated current task with new package {file.name}")
+                        else: raise DatabaseError
+                    except DatabaseError:
+                        session.cycle_key()
+                        create_upload_task(session)
```

Comment:
Twine uploads each package in a separate post request, but they all share the same session. To group the uploaded packages into one task I pass the Session object to the task and have it wait 5 seconds until starting the task. If more packages are uploaded in those 5 seconds then the session is updated with those tasks and the start time of the task is pushed back.  If a giant package is in the list of uploaded packages and the task starts before the package finishes uploading, then package's from that one and on will be added in a new task.